File: polyglot/csharp/root_jailbreak_detector.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Text;

namespace rootsentry
{
    /// <summary>
    /// Mobile runtime integrity detection with scored posture verdict.
    /// Detects root, jailbreak, emulator, hook, and tamper indicators.
    /// </summary>
    public static class RootJailbreakDetector
    {
        private const int MaxScore = 100;

        // Known root/jailbreak library signatures (partial names)
        private static readonly HashSet<string> RootLibSignatures = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "libroot", "libjailbreak", "libsu", "supersu", "magisk", 
            "zygote64_32", "zygote64_arm64_v8a", "arm64-v8a", "x86_64"
        };

        // Known emulator process names and env vars
        private static readonly HashSet<string> EmulatorProcessNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "com.android.emulator", "bluestacks", "nox", "mokee", 
            "genymotion", "android_studio_emu", "vysor"
        };

        // Common jailbreak environment variables
        private static readonly HashSet<string> JailbreakEnvVars = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "ANDROID_ROOT", "JAILBREAK_MODE", "CYDIA_HOME", 
            "DEBIAN_FRONTEND", "TERM_PROGRAM"
        };

        // Hooking library signatures
        private static readonly HashSet<string> HookLibSignatures = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "frida", "xposed", "lsposed", "cydia", "sileo", 
            "theos", "substrate", "hooker"
        };

        // Known tamper indicators (file paths that should exist in clean installs)
        private static readonly HashSet<string> SystemBinaries = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "/system/bin/sh", "/system/bin/zygote64", 
            "/data/local/tmp/frida-server-14.x.x"
        };

        public class PostureResult
        {
            public int Score { get; }
            public string Verdict { get; }
            public List<CheckDetail> Details { get; }
            
            public PostureResult(int score, string verdict, List<CheckDetail> details)
            {
                Score = score;
                Verdict = verdict;
                Details = details ?? new List<CheckDetail>();
            }

            public override string ToString() => $"{Verdict} (Score: {Score}/{MaxScore})";
        }

        /// <summary>
        /// Detailed result of a single integrity check.
        /// </summary>
        public class CheckDetail
        {
            public string Name { get; }
            public bool Passed { get; }
            public int PointsLost { get; }
            public string Message { get; }

            public CheckDetail(string name, bool passed, int pointsLost = 0, string message = null)
            {
                Name = name;
                Passed = passed;
                PointsLost = pointsLost;
                Message = message;
            }
        }

        /// <summary>
        /// Main entry point for integrity detection.
        /// </summary>
        public static PostureResult Detect()
        {
            var details = new List<CheckDetail>();
            int totalPointsLost = 0;

            // Step 1: Process and Assembly Integrity Checks
            details.AddRange(CheckProcessIntegrity());

            // Step 2: Environment Variable Analysis
            details.AddRange(CheckEnvironmentVariables());

            // Step 3: File System and Binary Verification
            details.AddRange(CheckFileSystemIntegrity());

            // Step 4: Memory/Loaded Library Analysis
            details.AddRange(CheckMemoryAndLibraries());

            // Calculate final score
            int totalPoints = MaxScore;
            totalPointsLost = details.Sum(d => d.PointsLost);
            int finalScore = Math.Max(0, totalPoints - totalPointsLost);

            string verdict;
            if (finalScore >= 90)
                verdict = "Clean";
            else if (finalScore >= 75)
                verdict = "Minor Suspicion";
            else if (finalScore >= 50)
                verdict = "Moderate Risk";
            else if (finalScore >= 25)
                verdict = "High Risk";
            else
                verdict = "Critical Compromise";

            return new PostureResult(finalScore, verdict, details);
        }

        /// <summary>
        /// Check process name, command line, and assembly location.
        /// </summary>
        private static IEnumerable<CheckDetail> CheckProcessIntegrity()
        {
            var details = new List<CheckDetail>();

            // Check current process name
            string processName = Process.GetCurrentProcess().ProcessName;
            
            if (EmulatorProcessNames.Contains(processName))
            {
                details.Add(new CheckDetail(
                    "Process Name", false, 15, 
                    $"Detected emulator-like process: {processName}"));
            }

            // Check command line arguments for suspicious flags
            string commandLine = Process.GetCurrentProcess().CommandLine;
            
            if (commandLine.Contains("-debug") || commandLine.Contains("--debug"))
            {
                details.Add(new CheckDetail(
                    "Debug Flags", false, 5, 
                    "Found debug flag in command line"));
            }

            // Check assembly location for tampering signs
            string assemblyLocation = Assembly.GetExecutingAssembly().Location;
            
            if (assemblyLocation.Contains("temp") || 
                assemblyLocation.Contains("cache") ||
                assemblyLocation.Contains("data"))
            {
                details.Add(new CheckDetail(
                    "Assembly Location", false, 10, 
                    $"Assembly in non-standard location: {assemblyLocation}"));
            }

            return details;
        }

        /// <summary>
        /// Analyze environment variables for jailbreak/emulator indicators.
        /// </summary>
        private static IEnumerable<CheckDetail> CheckEnvironmentVariables()
        {
            var details = new List<CheckDetail>();

            // Get all environment variables
            var envVars = Environment.GetEnvironmentVariables();

            foreach (var kvp in envVars)
            {
                string name = kvp.Key;
                
                if (JailbreakEnvVars.Contains(name))
                {
                    details.Add(new CheckDetail(
                        $"Env Var: {name}", false, 10, 
                        $"Found jailbreak indicator: {name}"));
                }

                // Check for emulator-specific env vars
                if (name.StartsWith("EMULATOR") || name.Contains("_EMU_"))
                {
                    details.Add(new CheckDetail(
                        $"Emulator Env Var", false, 8, 
                        $"Found emulator env var: {name}"));
                }

                // Check for common tamper indicators
                if (name == "USER" && !kvp.Value.ToString().Contains("root"))
                {
                    details.Add(new CheckDetail(
                        "User Context", true, 0, 
                        $"Running as user: {kvp.Value}"));
                }
            }

            return details;
        }

        /// <summary>
        /// Verify file system integrity and check for modified binaries.
        /// </summary>
        private static IEnumerable<CheckDetail> CheckFileSystemIntegrity()
        {
            var details = new List<CheckDetail>();

            // Check for root/jailbreak files in common locations
            string[] rootIndicators = {
                "/data/local/tmp/frida-server",
                "/data/data/com.termux/files/usr/bin/root",
                "/system/etc/init/android_root_init.sh"
            };

            foreach (var indicator in rootIndicators)
            {
                if (!string.IsNullOrEmpty(indicator) && 
                    File.Exists(indicator))
                {
                    details.Add(new CheckDetail(
                        $"Root Indicator: {indicator}", false, 12, 
                        $"Found potential root file at: {indicator}"));
                }
            }

            // Check for modified system binaries (if accessible)
            try
            {
                string[] systemBins = { "/system/bin/sh", "/system/bin/zygote64" };
                
                foreach (var bin in systemBins)
                {
                    if (!string.IsNullOrEmpty(bin))
                    {
                        if (File.Exists(bin))
                        {
                            // Check file size for anomalies
                            long fileSize = new FileInfo(bin).Length;
                            
                            // Typical sizes vary, but extremely large files are suspicious
                            if (fileSize > 50 * 1024 * 1024) // 50MB threshold
                            {
                                details.Add(new CheckDetail(
                                    $"Binary Size: {bin}", false, 8, 
                                    $"Unusually large binary file: {fileSize} bytes"));
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                // Expected in many mobile environments where /system is read-only
                details.Add(new CheckDetail(
                    "System Binary Access", true, 0, 
                    $"Limited system access: {ex.Message}"));
            }

            return details;
        }

        /// <summary>
        /// Analyze loaded libraries and memory for hooking indicators.
        /// </summary>
        private static IEnumerable<CheckDetail> CheckMemoryAndLibraries()
        {
            var details = new List<CheckDetail>();

            // Get all loaded assemblies (DLLs)
            var loadedAssemblies = AppDomain.CurrentDomain.GetAssemblies();

            foreach (var assembly in loadedAssemblies)
            {
                string name = assembly.GetName().Name;
                
                if (RootLibSignatures.Contains(name))
                {
                    details.Add(new CheckDetail(
                        $"Loaded Library: {name}", false, 15, 
                        $"Found root/jailbreak library: {name}"));
                }

                if (HookLibSignatures.Contains(name))
                {
                    details.Add(new CheckDetail(
                        $"Hooking Library: {name}", false, 20, 
                        $"Found hooking library: {name}"));
                }

                // Check for suspicious assembly locations
                string location = assembly.Location;
                
                if (location != null && 
                    (location.Contains("temp") || 
                     location.Contains("/data/data/") && !location.Contains(".apk")))
                {
                    details.Add(new CheckDetail(
                        $"Assembly Location: {name}", false, 5, 
                        $"Loaded from non-standard path: {location}"));
                }
            }

            // Check for Frida server specifically (common hooking tool)
            try
            {
                string fridaPath = "/data/local/tmp/frida-server-14.x.x";
                
                if (!string.IsNullOrEmpty(fridaPath))
                {
                    if (File.Exists(fridaPath))
                    {
                        details.Add(new CheckDetail(
                            "Frida Server", false, 25, 
                            $"Found Frida server at: {fridaPath}"));
                    }
                }
            }
            catch
            {
                // Expected in many environments
            }

            return details;
        }

        /// <summary>
        /// Get a human-readable interpretation of the score.
        /// </summary>
        public static string InterpretScore(int score)
        {
            switch (score)
            {
                case >= 90:
                    return "Runtime appears clean and trustworthy.";
                case >= 75:
                    return "Minor indicators present, likely safe but worth noting.";
                case >= 50:
                    return "Moderate risk detected. Review details for context.";
                case >= 25:
                    return "High probability of runtime modification or emulation.";
                default:
                    return "Strong indicators of root/jailbreak/emulator environment.";
            }
        }

        /// <summary>
        /// Main demo/entry point. Run this to test the detector.
        /// </summary>
        public static void Main(string[] args)
        {
            Console.WriteLine("=== rootsentry: Mobile Runtime Integrity Detector ===\n");

            // Run detection
            var result = Detect();

            // Display results
            Console.WriteLine($"Verdict: {result.Verdict}");
            Console.WriteLine($"Score:   {result.Score}/{MaxScore}\n");
            
            if (string.IsNullOrEmpty(result.Interpretation()))
                Console.WriteLine(InterpretScore(result.Score));
            else
                Console.WriteLine(result.Interpretation());

            Console.WriteLine("\n--- Detailed Findings ---\n");

            // Sort by points lost (worst first)
            var sortedDetails = result.Details.OrderByDescending(d => d.PointsLost).ToList();

            foreach (var detail in sortedDetails)
            {
                string status = detail.Passed ? "[PASS]" : "[FAIL]";
                Console.WriteLine($"{status} | {detail.Name,-35}");
                
                if (!detail.Passed)
                {
                    Console.WriteLine($"       Points lost: {detail.PointsLost}");
                    if (!string.IsNullOrEmpty(detail.Message))
                        Console.WriteLine($"       Details: {detail.Message}");
                }
            }

            // Summary
            int failedCount = result.Details.Count(d => !d.Passed);
            Console.WriteLine($"\n--- Summary ---");
            Console.WriteLine($"Total checks:  {result.Details.Count}");
            Console.WriteLine($"Passed:        {result.Details.Count(d => d.Passed)}");
            Console.WriteLine($"Failed:        {failedCount}");
            Console.WriteLine($"Score impact:  -{result.Details.Sum(d => d.PointsLost)}/{MaxScore} points");

            // Exit with appropriate code
            Environment.Exit(result.Score < 50 ? 1 : 0);
        }

        /// <summary>
        /// Get human-readable interpretation of the result.
        /// </summary>
        public string Interpretation() => InterpretScore(Score);
    }
}

File: polyglot/rust/root_jailbreak_detector.rs
mod root_jailbreak_detector {
    use std::collections::{HashMap, HashSet};
    use std::env;
    use std::fs;
    use std::io::{self, Read, Write};
    use std::path::PathBuf;
    use std::process::{Command, Stdio};

    // =====================================================================
    // CONSTANTS & CONFIGURATION
    // =====================================================================

    /// Known root/jailbreak indicator paths (Android/iOS)
    const ROOT_INDICATORS: &[&str] = &[
        "/data/local/tmp/su",
        "/system/bin/sh",
        "/system/xbin/which",
        "/system/app/SuperUser.apk",
        "/system/usr/cmdline.txt",
        "/proc/self/status",
    ];

    /// Known emulator indicators
    const EMULATOR_INDICATORS: &[&str] = &[
        "/data/local/tmp/emulator",
        "/dev/socket/qemmd",
        "/data/data/com.android.vending/backup_set",
        "/system/etc/init.qemu.sh.rc",
        "QEMU_AUDIO_DRV=none",
    ];

    /// Known hook/tamper indicators
    const HOOK_INDICATORS: &[&str] = &[
        "/data/local/tmp/hooks.so",
        "/system/lib/libhooker.so",
        "/proc/1/cmdline",
        "LD_PRELOAD=",
        "DYLD_INSERT_LIBRARIES=",
    ];

    // =====================================================================
    // TYPES & ENUMS
    // =====================================================================

    /// Result of a single integrity check
    #[derive(Debug, Clone)]
    pub struct CheckResult {
        name: String,
        passed: bool,
        score_impact: u8,
        message: Option<String>,
    }

    impl Default for CheckResult {
        fn default() -> Self {
            Self {
                name: "unknown".to_string(),
                passed: true,
                score_impact: 0,
                message: None,
            }
        }
    }

    /// Severity level of a detected issue
    #[derive(Debug, Clone, Copy, PartialEq)]
    pub enum Severity {
    Low(1),
    Medium(2),
    High(3),
    Critical(4),
}

impl Default for Severity {
    fn default() -> Self {
        Self::Low
    }
}

/// Final posture verdict with remediation guidance
#[derive(Debug, Clone)]
pub struct PostureVerdict {
    pub score: u8,
    pub level: Level,
    pub findings: Vec<Finding>,
    pub recommendation: String,
}

impl Default for PostureVerdict {
    fn default() -> Self {
        Self {
            score: 0,
            level: Level::Clean,
            findings: Vec::new(),
            recommendation: "Application appears to be running in a trusted environment.".to_string(),
        }
    }
}

/// Severity of the overall posture
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Level {
    Clean(0),
    LowRisk(1..=25),
    MediumRisk(26..=50),
    HighRisk(51..=75),
    Critical(76..=100),
}

impl Default for Level {
    fn default() -> Self {
        Self::Clean
    }
}

/// A single finding with context
#[derive(Debug, Clone)]
pub struct Finding {
    pub category: Category,
    pub severity: Severity,
    pub source: String,
    pub detail: String,
}

impl Default for Finding {
    fn default() -> Self {
        Self {
            category: Category::Unknown,
            severity: Severity::Low,
            source: "unknown".to_string(),
            detail: String::new(),
        }
    }
}

/// Category of the finding
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Category {
    FileSystem,
    Process,
    Environment,
    Network,
    Memory,
    Unknown,
}

impl Default for Category {
    fn default() -> Self {
        Self::Unknown
    }
}

// =====================================================================
    // CORE DETECTION ENGINE
    // =====================================================================

    /// Main detector instance
    pub struct RootJailbreakDetector {
        config: Config,
        results: Vec<CheckResult>,
    }

    impl Default for RootJailbreakDetector {
        fn default() -> Self {
            Self::new(Config::default())
        }
    }

    /// Detector configuration
    #[derive(Debug, Clone)]
    pub struct Config {
        pub strict_mode: bool,
        pub include_network: bool,
        pub verbose: bool,
    }

    impl Default for Config {
        fn default() -> Self {
            Self {
                strict_mode: false,
                include_network: true,
                verbose: false,
            }
        }
    }

    impl RootJailbreakDetector {
        pub fn new(config: Config) -> Self {
            Self {
                config,
                results: Vec::new(),
            }
        }

        /// Run all integrity checks and return the posture verdict
        pub fn detect(&mut self) -> PostureVerdict {
            self.results.clear();

            // File system checks
            for indicator in ROOT_INDICATORS.iter() {
                if let Some(result) = Self::check_file_path(indicator, "root_indicator") {
                    self.results.push(result);
                }
            }

            // Emulator checks
            for indicator in EMULATOR_INDICATORS.iter() {
                if let Some(result) = Self::check_file_path(indicator, "emulator_indicator") {
                    self.results.push(result);
                }
            }

            // Hook/tamper checks
            for indicator in HOOK_INDICATORS.iter() {
                if let Some(result) = Self::check_file_path(indicator, "hook_indicator") {
                    self.results.push(result);
                }
            }

            // Environment variable checks
            Self::check_environment_variables(&mut self.results);

            // Process/proc filesystem checks (Linux-specific)
            #[cfg(target_os = "linux")]
            Self::check_proc_self_status();

            // Calculate final score and verdict
            let total_score: u8 = self.results.iter().map(|r| r.score_impact).sum();
            let max_possible_score = 100;
            let normalized_score = (total_score as f32 / max_possible_score as f32) * 100u8 as f32 as u8;

            PostureVerdict {
                score: normalized_score,
                level: Level::from_score(normalized_score),
                findings: self.results.iter()
                    .filter(|r| !r.passed)
                    .map(|r| Finding {
                        category: Category::FileSystem,
                        severity: Severity::High,
                        source: r.name.clone(),
                        detail: r.message.clone().unwrap_or_default(),
                    })
                    .collect(),
                recommendation: self.generate_recommendation(normalized_score),
            }
        }

        /// Check if a file path exists and is suspicious
        fn check_file_path(path: &str, category: &str) -> Option<CheckResult> {
            let full_path = PathBuf::from(path);
            
            if !full_path.exists() {
                return None;
            }

            // Check if it's a symlink (common in root/jailbreak scenarios)
            let is_symlink = match fs::read_link(&full_path) {
                Ok(_) => true,
                Err(_) => false,
            };

            Some(CheckResult {
                name: format!("{} at {}", category, path),
                passed: !is_symlink,
                score_impact: if is_symlink { 15 } else { 0 },
                message: Some(format!(
                    "Suspicious symlink detected at: {}",
                    full_path.display()
                )),
            })
        }

        /// Check environment variables for tampering indicators
        fn check_environment_variables(results: &mut Vec<CheckResult>) {
            let suspicious_vars = [
                ("LD_PRELOAD", "LD_PRELOAD"),
                ("DYLD_INSERT_LIBRARIES", "DYLD_INSERT_LIBRARIES"),
                ("ANDROID_DEBUG", "ANDROID_DEBUG"),
                ("QEMU_AUDIO_DRV", "QEMU_AUDIO_DRV"),
            ];

            for (var, name) in suspicious_vars.iter() {
                if let Ok(value) = env::var(var) {
                    results.push(CheckResult {
                        name: format!("{}={}", name, value),
                        passed: value.is_empty(),
                        score_impact: 10,
                        message: Some(format!(
                            "Suspicious environment variable set: {}={}",
                            name, value
                        )),
                    });
                }
            }
        }

        /// Check /proc/self/status for process anomalies (Linux)
        fn check_proc_self_status() {
            if let Ok(status) = fs::read_to_string("/proc/self/status") {
                // Look for common indicators in status file
                let lines: Vec<&str> = status.lines().collect();
                
                for line in lines.iter() {
                    if line.starts_with("VmRSS:") || line.starts_with("Threads:") {
                        // Normal fields, continue checking
                    } else if line.contains("state") && !line.contains("S") {
                        // Non-idle state might indicate activity
                    }
                }
            }
        }

        /// Generate recommendation based on score
        fn generate_recommendation(score: u8) -> String {
            match Level::from_score(score) {
                Level::Clean => "Application appears to be running in a trusted environment.".to_string(),
                Level::LowRisk => "Minor indicators detected. Review findings and consider additional verification if strict mode is enabled.".to_string(),
                Level::MediumRisk => "Moderate risk detected. Recommend investigation of flagged paths and potential re-deployment from verified build artifacts.".to_string(),
                Level::HighRisk => "Significant tampering indicators found. Verify application integrity, check for active hooks, and consider emergency patch deployment.".to_string(),
                Level::Critical => "Critical compromise likely. Immediate forensic analysis recommended. Check for root access, active emulators, or injected code.".to_string(),
            }
        }

        /// Convert score to level
        fn from_score(score: u8) -> Level {
            match score {
                0..=25 => Level::LowRisk,
                26..=50 => Level::MediumRisk,
                51..=75 => Level::HighRisk,
                _ => Level::Critical,
            }
        }

        /// Get human-readable level name
        pub fn level_name(level: &Level) -> &'static str {
            match level {
                Level::Clean(_) => "CLEAN",
                Level::LowRisk(_) => "LOW RISK",
                Level::MediumRisk(_) => "MEDIUM RISK",
                Level::HighRisk(_) => "HIGH RISK",
                Level::Critical(_) => "CRITICAL",
            }
        }

        /// Get detailed findings as JSON-like string
        pub fn get_findings_json(&self) -> String {
            self.results.iter()
                .map(|r| format!(
                    r#"{{"name":"{}","passed":{},,"score_impact":{}}}"#,
                    r.name, r.passed, r.score_impact
                ))
                .collect::<Vec<_>>()
                .join(", ")
        }

        /// Check a single path and return result (for incremental checking)
        pub fn check_path(&self, path: &str) -> Option<CheckResult> {
            Self::check_file_path(path, "manual_check")
        }

        /// Get the number of failed checks
        pub fn failed_count(&self) -> usize {
            self.results.iter().filter(|r| !r.passed).count()
        }

        /// Get total score impact
        pub fn total_score_impact(&self) -> u8 {
            self.results.iter().map(|r| r.score_impact).sum()
        }
    }

    // =====================================================================
    // PUBLIC API - CONVENIENCE FUNCTIONS
    // =====================================================================

    /// Quick check with default configuration
    pub fn quick_check() -> PostureVerdict {
        let mut detector = RootJailbreakDetector::default();
        detector.detect()
    }

    /// Check a specific path for root indicators
    pub fn check_root_indicator(path: &str) -> Result<bool, io::Error> {
        Ok(fs::metadata(path)?.is_symlink())
    }

    /// Get the current posture score without running full detection
    pub fn get_current_score() -> u8 {
        let mut detector = RootJailbreakDetector::default();
        let verdict = detector.detect();
        verdict.score
    }

    // =====================================================================
    // DEMO / ENTRY POINT
    // =====================================================================

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn test_default_detector() {
            let mut detector = RootJailbreakDetector::default();
            let verdict = detector.detect();
            
            assert!(verdict.score <= 100);
            assert!(verdict.level != Level::Critical || verdict.score > 75);
        }

        #[test]
        fn test_file_check() {
            // Check a known existing path
            let result = RootJailbreakDetector::check_file_path("/etc/passwd", "test");
            assert!(result.is_some());
            
            // Check non-existent path
            let result = RootJailbreakDetector::check_file_path("/nonexistent/path/12345", "test");
            assert!(result.is_none());
        }

        #[test]
        fn test_environment_check() {
            env::set_var("TEST_VAR", "test_value");
            
            let mut detector = RootJailbreakDetector::default();
            let _verdict = detector.detect();
        }
    }

    /// Main demo function - run when compiled as binary
    fn main_demo() -> Result<(), Box<dyn std::error::Error>> {
        println!("=== Rootsentry: Mobile Runtime Integrity Detector ===\n");

        // Run detection
        let mut detector = RootJailbreakDetector::default();
        let verdict = detector.detect();

        // Output results
        println!("Posture Verdict: {}", Level::level_name(&verdict.level));
        println!("Risk Score: {}/100", verdict.score);
        println!("\nRecommendation:\n  {}\n", verdict.recommendation);

        if !verdict.findings.is_empty() {
            println!("\n=== Findings ({}) ===", verdict.findings.len());
            for finding in &verdict.findings {
                println!("  [{}] {}", 
                    Severity::level_name(finding.severity),
                    finding.source,
                );
                if !finding.detail.is_empty() {
                    println!("      Detail: {}", finding.detail);
                }
            }
        }

        // Show detailed results for debugging
        println!("\n=== Detailed Check Results ===");
        for (i, result) in detector.results.iter().enumerate() {
            let status = if result.passed { "✓" } else { "✗" };
            println!("  [{:2}] {} - {}", i + 1, status, result.name);
            if !result.message.is_some_and(|m| m.is_empty()) {
                println!("         {}", result.message.as_ref().unwrap());
            }
        }

        Ok(())
    }

    /// Helper to get level name for display
    impl Level {
        pub fn level_name(&self) -> &'static str {
            match self {
                Self::Clean(_) => "CLEAN",
                Self::LowRisk(_) => "LOW RISK",
                Self::MediumRisk(_) => "MEDIUM RISK",
                Self::HighRisk(_) => "HIGH RISK",
                Self::Critical(_) => "CRITICAL",
            }
        }
    }

    /// Helper to get severity name for display
    impl Severity {
        pub fn level_name(&self) -> &'static str {
            match self {
                Self::Low(1) => "LOW",
                Self::Medium(2) =>